openverse_catalog/dags/common/sensors/single_run_external_dags_sensor.py

```diff
@@ -0,0 +1,88 @@
+import os
+from typing import Iterable
+
+from airflow.exceptions import AirflowException
+from airflow.models import DagBag, DagModel, DagRun, TaskInstance
+from airflow.sensors.base import BaseSensorOperator
+from airflow.utils.session import provide_session
+from airflow.utils.state import State
+
+
+class SingleRunExternalDAGsSensor(BaseSensorOperator):
+    """
+    Waits for a list of related DAGs, each assumed to have a similar Sensor,
+    to not be running. A related DAG is considered to be 'running' if it is
+    itself in the running state, and its corresponding wait task completed
+    successfully. It looks for a task with the same `task_id` in each of the
+    provided external DAGs.
+
+    :param external_dag_ids: A list of dag_ids that you want to wait for
+    :param check_existence: Set to `True` to check if the external DAGs exist,
+    and immediately cease waiting if not (default value: False).
+    """
+
+    def __init__(
+        self,
+        *,
+        external_dag_ids: Iterable[str],
+        check_existence: bool = False,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.external_dag_ids = external_dag_ids
+        self.check_existence = check_existence
+        self._has_checked_existence = False
+
+    @provide_session
+    def poke(self, context, session=None):
+        self.log.info(
+            "Poking for DAGs %s ...",
+            self.external_dag_ids,
+        )
+
+        # Check DAG existence only once, on the first execution.
+        if self.check_existence and not self._has_checked_existence:
+            self._check_for_existence(session=session)
```

Comment:
nit (and a very minor one at that) about readability for `poke`: I'm not sure if this is just me, but setting `_has_checked_existence` inside of `_check_for_existence` is slightly confusing when reading the overall flow of the `poke` method, unless you dig into `check_for_existence` while reading through `poke`. But it's sort of like a side effect that seems to belong more to `poke` given `poke` itself is the thing commanding the checking :thinking: I suppose if the conditional was moved all into `check_for_existence` (as an early return conditional) then it would make more sense and would reduce the scope of `poke`; then `poke` doesn't concern itself with short-circuiting the check and lets `check_for_existence` handle all the existence checking logic including the short-circuiting.